lib/kubernetes-deploy/template_sets.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module KubernetesDeploy
+  class TemplateSets
+    VALID_TEMPLATES = %w(.yml.erb .yml .yaml .yaml.erb)
+    class << self
+      def new_from_dirs_and_files(paths:, logger:, current_sha:, bindings:)
+        resource_templates = {}
+        dir_paths, file_paths = paths.partition { |path| File.directory?(path) }
+
+        # Directory paths
+        dir_paths.each do |template_dir|
+          resource_templates[template_dir] = Dir.foreach(template_dir).select do |filename|
+            filename.end_with?(*VALID_TEMPLATES) || filename == EjsonSecretProvisioner::EJSON_SECRETS_FILE
+          end
+        end
+
+        # Filename paths
+        file_paths.each do |filename|
+          dir_name = File.dirname(filename)
+          resource_templates[dir_name] ||= []
+          resource_templates[dir_name] << File.basename(filename) unless resource_templates[dir_name].include?(filename)
+        end
+
+        template_sets = TemplateSets.new
+        resource_templates.map do |template_dir, files|
+          template_sets << TemplateSet.new(template_dir: template_dir, file_whitelist: files, logger: logger,
+              renderer: Renderer.new(
+                current_sha: current_sha,
+                logger: logger,
+                bindings: bindings,
+                template_dir: template_dir
+              ))
+        end
+        template_sets
+      end
+    end
+
+    def <<(template_set)
+      if template_set.is_a?(TemplateSet)
+        @template_sets << template_set
+      else
+        raise InvalidTemplateError, "Expected TemplateSet but got #{template_set.class}"
+      end
+    end
+
+    def with_resource_definitions(render_erb: false)
+      @template_sets.flat_map do |template_set|
+        template_set.with_resource_definitions(render_erb: render_erb) do |r_def|
+          yield r_def
+        end
+      end
+    end
+
+    def ejson_secrets_files
+      @template_sets.map(&:ejson_secrets_file).compact
+    end
+
+    def validate
+      errors = []
+      @template_sets.each do |template_set|
+        errors << template_set.validate
+      end
+      errors.flatten
+    end
+
+    private
+
+    def initialize
+      @template_sets = []
+    end
+  end
+
+  class TemplateSet
```

Comment:
Ideally this would be a private inner class but that's not really something you can do in Ruby. Scoping it as `TemplateSets::TemplateSet` also felt wrong. I don't think this should live in its own file, though, since it's intrinsically linked to `TemplateSets` and it's the latter that we should expose and make visible to the rest of the library